File: benchmarks/hypergeometric-pfq/golden/verify.ts
// =============================================================================
// benchmarks/hypergeometric-pfq/golden/verify.ts — invariant verifier (TS port)
// =============================================================================
//
// stdin:
//   { input:     <input row from inputs.json>,
//     candidate: <candidate output — see PROMPT.md for shapes>,
//     id:        string }
//
// stdout:
//   { pass: bool, reason: string,
//     checks: { <name>: { pass: bool, detail: string } } }
//
// This is the TypeScript port of verify.py per ADR-0028 §4.
// Tolerances are preserved byte-for-byte with verify.py.
// Do NOT tighten or loosen during migration.
//
// The verifier is self-contained: value-accuracy comparisons use
// @workbench/bigfloat (arbitrary-precision BigInt substrate, ADR-0020)
// imported dynamically from the workbench repo via WORKBENCH_ROOT.
// No mpmath / Python runtime is required at grading time.
//
// ─── Checks ──────────────────────────────────────────────────────────────────
//
//   no_tool_error            — never admissible; always checked first.
//   shape                    — success-path: required fields, types.
//   finite_value             — success-path: re/im parse as finite BigFloat.
//   method_admissible        — success-path: method ∈ admitted set.
//   self_reported_precision  — success-path: 0 ≤ achieved_precision ≤ requested.
//   value_accuracy           — success-path: rel-err vs pinned truth ≤ tolerance_rel.
//   boundary_envelope        — refusal-path: kind=tagged, tag matches, payload predicate.
//
// Determinism: BigInt-substrate arithmetic (ADR-0020 arbprec tier) is
// bit-identical cross-platform forever.  The expected.json truth values are
// pinned at ~80 dps; comparisons at 128 dps (≈385 bits) clear every case's
// tolerance by ≥ 20 orders of magnitude.
//
// =============================================================================

import { readFileSync } from "node:fs";
import { resolve } from "node:path";

// ─── Resolve workbench root ───────────────────────────────────────────────────
//
// The corpus grader sets WORKBENCH_ROOT when it spawns this verifier.
// The default fallback assumes the conventional side-by-side layout:
//   ~/Projects/scientist-workbench/         ← WORKBENCH_ROOT
//   ~/Projects/scientist-workbench-corpus/  ← this repo

const WORKBENCH_ROOT: string =
  process.env["WORKBENCH_ROOT"] ??
  resolve(import.meta.dir, "..", "..", "..", "..", "scientist-workbench");

// ─── Dynamic imports (resolved through workbench) ────────────────────────────

const bigfloatPath = resolve(WORKBENCH_ROOT, "packages/bigfloat/src/index.ts");
// We import at runtime from the workbench path, typed structurally.
type BfModule = {
  decimalToBinaryPrecision: (dps: number) => number;
  fromString: (s: string, prec: number) => unknown;
  add: (a: unknown, b: unknown, prec: number) => unknown;
  sub: (a: unknown, b: unknown, prec: number) => unknown;
  mul: (a: unknown, b: unknown, prec: number) => unknown;
  div: (a: unknown, b: unknown, prec: number) => unknown;
  sqrt: (a: unknown, prec: number) => unknown;
  le: (a: unknown, b: unknown) => boolean;
  isZero: (a: unknown) => boolean;
  toFloat64: (a: unknown) => { value: number; overflow: null | string };
};
const bf = await import(bigfloatPath) as BfModule;

// BigFloat comparison precision: 128 decimal digits = ~425 binary bits.
// Large enough to distinguish any pair of truth/candidate values the bench uses
// and to clear the loosest tolerance (1e-40) by many orders of magnitude.
const VERIFY_DPS = 128;
const VERIFY_PREC = bf.decimalToBinaryPrecision(VERIFY_DPS);

// ─── Check type ───────────────────────────────────────────────────────────────

type Check = { pass: boolean; detail: string };

// ─── expected.json loader ────────────────────────────────────────────────────

type ExpectedCase = {
  id: string;
  tier: string;
  tolerance_rel: string;
  expected: {
    kind: "value" | "tagged";
    truth?: { re: string; im: string };
    tag?: string;
    payload_predicate?: { reason_substr?: string };
  };
};

type ExpectedIndex = Map<string, ExpectedCase>;

let _expectedIndex: ExpectedIndex | null = null;

function loadExpected(): ExpectedIndex {
  if (_expectedIndex !== null) return _expectedIndex;
  const path = resolve(import.meta.dir, "expected.json");
  const payload = JSON.parse(readFileSync(path, "utf8")) as {
    cases: ExpectedCase[];
  };
  _expectedIndex = new Map(payload.cases.map((c) => [c.id, c]));
  return _expectedIndex;
}

// ─── Per-check helpers ────────────────────────────────────────────────────────

const ADMITTED_METHODS = new Set([
  "closed-form-0F0",
  "closed-form-1F0",
  "direct-series",
]);

function checkShape(candidate: Record<string, unknown>): Check {
  // Required success-record fields:
  //   value: { re: string, im: string }
  //   achieved_precision: number (int)
  //   method: string
  //   n_terms: number (int)
  //   working_precision: number (int)
  //   warnings: string[]
  const required = [
    "value",
    "achieved_precision",
    "method",
    "n_terms",
    "working_precision",
    "warnings",
  ];
  const missing = required.filter((k) => !(k in candidate));
  if (missing.length > 0) {
    return { pass: false, detail: `missing fields: ${JSON.stringify(missing.sort())}` };
  }

  const v = candidate["value"];
  if (
    typeof v !== "object" ||
    v === null ||
    Array.isArray(v) ||
    !("re" in v) ||
    !("im" in v)
  ) {
    return {
      pass: false,
      detail: `value must be {re, im}; got ${typeof v === "object" ? JSON.stringify(v) : typeof v}`,
    };
  }
  const vObj = v as Record<string, unknown>;
  if (typeof vObj["re"] !== "string" || typeof vObj["im"] !== "string") {
    return { pass: false, detail: "value.re and value.im must be strings" };
  }
  if (typeof candidate["achieved_precision"] !== "number") {
    return { pass: false, detail: "achieved_precision must be int" };
  }
  if (typeof candidate["n_terms"] !== "number") {
    return { pass: false, detail: "n_terms must be int" };
  }
  if (typeof candidate["working_precision"] !== "number") {
    return { pass: false, detail: "working_precision must be int" };
  }
  if (typeof candidate["method"] !== "string") {
    return { pass: false, detail: "method must be str" };
  }
  const warnings = candidate["warnings"];
  if (
    !Array.isArray(warnings) ||
    !(warnings as unknown[]).every((w) => typeof w === "string")
  ) {
    return { pass: false, detail: "warnings must be list[str]" };
  }
  return { pass: true, detail: "all required fields present" };
}

function checkFiniteValue(candidate: Record<string, unknown>): Check {
  // Parse value.re and value.im as BigFloat at VERIFY_PREC.
  // A finite value parses without exception and is not ±∞ (bigfloat
  // doesn't represent infinities; a parse failure is caught as thrown).
  const v = candidate["value"] as { re: string; im: string };
  try {
    const re = bf.fromString(v.re, VERIFY_PREC);
    const im = bf.fromString(v.im, VERIFY_PREC);
    // BigFloat has no ±∞ representation — any successful fromString is finite.
    void re;
    void im;
  } catch (e) {
    return { pass: false, detail: `could not parse value: ${(e as Error).message}` };
  }
  return { pass: true, detail: "value parses, finite" };
}

function checkMethodAdmissible(candidate: Record<string, unknown>): Check {
  const m = candidate["method"] as string;
  if (!ADMITTED_METHODS.has(m)) {
    return {
      pass: false,
      detail: `method=${JSON.stringify(m)} not in admitted set ${JSON.stringify([...ADMITTED_METHODS].sort())}`,
    };
  }
  return { pass: true, detail: `method=${JSON.stringify(m)}` };
}

function checkSelfReportedPrecision(
  candidate: Record<string, unknown>,
  requestedPrecision: number,
): Check {
  // Per verify.py: 0 ≤ achieved_precision ≤ requested_precision.
  // Under-reporting is permitted (graceful degradation); over-reporting is not.
  const ap = candidate["achieved_precision"] as number;
  if (ap < 0) {
    return { pass: false, detail: `achieved_precision=${ap} is negative` };
  }
  if (ap > requestedPrecision) {
    return {
      pass: false,
      detail: `achieved_precision=${ap} exceeds requested precision ${requestedPrecision} — tool over-reports`,
    };
  }
  return {
    pass: true,
    detail: `achieved=${ap}, requested=${requestedPrecision}`,
  };
}

function checkValueAccuracy(
  candidate: Record<string, unknown>,
  caseExpected: ExpectedCase,
): Check {
  // Compare candidate value to pinned truth using BigFloat arithmetic at
  // VERIFY_PREC (128 dps) — the same mathematical operation as verify.py's
  // mpmath comparison at 80 dps, but with a wider precision budget.
  //
  // rel_err = |candidate - truth| / |truth|      (fallback to absolute if truth=0)
  //
  // pass iff rel_err ≤ tolerance_rel (parsed as a BigFloat at VERIFY_PREC)
  const truth = caseExpected.expected.truth!;
  const v = candidate["value"] as { re: string; im: string };

  // Parse everything at VERIFY_PREC.
  const truthRe = bf.fromString(truth.re, VERIFY_PREC);
  const truthIm = bf.fromString(truth.im, VERIFY_PREC);
  const candRe = bf.fromString(v.re, VERIFY_PREC);
  const candIm = bf.fromString(v.im, VERIFY_PREC);

  // diff = candidate - truth  (as complex)
  const diffRe = bf.sub(candRe, truthRe, VERIFY_PREC);
  const diffIm = bf.sub(candIm, truthIm, VERIFY_PREC);

  // |diff| = sqrt(diffRe² + diffIm²)
  const diffAbsSq = bf.add(
    bf.mul(diffRe, diffRe, VERIFY_PREC),
    bf.mul(diffIm, diffIm, VERIFY_PREC),
    VERIFY_PREC,
  );
  const diffAbs = bf.sqrt(diffAbsSq, VERIFY_PREC);

  // |truth| = sqrt(truthRe² + truthIm²)
  const truthAbsSq = bf.add(
    bf.mul(truthRe, truthRe, VERIFY_PREC),
    bf.mul(truthIm, truthIm, VERIFY_PREC),
    VERIFY_PREC,
  );
  const truthAbs = bf.sqrt(truthAbsSq, VERIFY_PREC);

  let rel: ReturnType<typeof bf.fromString>;
  if (bf.isZero(truthAbs)) {
    // Truth is zero — fall back to absolute error.
    rel = diffAbs;
  } else {
    rel = bf.div(diffAbs, truthAbs, VERIFY_PREC);
  }

  // Parse tolerance_rel.  The pinned values are like "1e-48", "1e-40".
  // bf.fromString handles scientific notation.
  const tol = bf.fromString(caseExpected.tolerance_rel, VERIFY_PREC);

  // rel ≤ tol  iff  cmp(rel, tol) ≤ 0
  const pass = bf.le(rel, tol);

  // Format for detail message using float64 approximations (fine for display).
  // bf.toFloat64 returns { value: number, overflow: ... } — extract .value.
  const relFloat = bf.toFloat64(rel).value;
  const tolFloat = bf.toFloat64(tol).value;

  if (pass) {
    return {
      pass: true,
      detail: `rel=${relFloat.toExponential(3)} ≤ tol=${tolFloat.toExponential(3)}`,
    };
  }
  return {
    pass: false,
    detail: `rel=${relFloat.toExponential(3)} > tol=${tolFloat.toExponential(3)}`,
  };
}

function checkBoundaryEnvelope(
  candidate: unknown,
  caseExpected: ExpectedCase,
): Check {
  // The candidate must be { kind:"tagged", tag:"hypergeometric-pfq/<class>", payload:{...} }.
  if (
    typeof candidate !== "object" ||
    candidate === null ||
    (candidate as Record<string, unknown>)["kind"] !== "tagged"
  ) {
    const got =
      typeof candidate === "object" && candidate !== null
        ? ((candidate as Record<string, unknown>)["kind"] ?? typeof candidate)
        : typeof candidate;
    return {
      pass: false,
      detail: `expected tagged refusal but got ${got}`,
    };
  }
  const c = candidate as Record<string, unknown>;
  const expectedTag = caseExpected.expected.tag!;
  const actualTag = (c["tag"] as string) ?? "";
  if (actualTag !== expectedTag) {
    return {
      pass: false,
      detail: `tag ${JSON.stringify(actualTag)} != expected ${JSON.stringify(expectedTag)}`,
    };
  }

  // Optional payload predicate: reason_substr
  const pred = caseExpected.expected.payload_predicate;
  if (pred?.reason_substr) {
    const payload = c["payload"];
    const reason =
      typeof payload === "object" && payload !== null
        ? ((payload as Record<string, unknown>)["reason"] as string) ?? ""
        : "";
    if (!reason.includes(pred.reason_substr)) {
      return {
        pass: false,
        detail: `payload.reason=${JSON.stringify(reason)} does not contain the predicate substring ${JSON.stringify(pred.reason_substr)}`,
      };
    }
  }

  return { pass: true, detail: `tagged ${expectedTag}` };
}

// ─── Wrap helper ─────────────────────────────────────────────────────────────

function wrap(checks: Record<string, Check>): {
  pass: boolean;
  reason: string;
  checks: Record<string, Check>;
} {
  const overall = Object.values(checks).every((c) => c.pass);
  if (overall) {
    return { pass: true, reason: "all invariants hold", checks };
  }
  const firstFail = Object.entries(checks).find(([, v]) => !v.pass)!;
  return {
    pass: false,
    reason: `failed: ${firstFail[0]} — ${firstFail[1].detail}`,
    checks,
  };
}

// ─── Top-level verify ─────────────────────────────────────────────────────────

function verify(payload: {
  id?: string;
  input?: Record<string, unknown>;
  candidate?: unknown;
}) {
  const caseId = payload["id"] ?? "";
  const candidate = payload["candidate"] ?? {};
  const inp = payload["input"] ?? {};

  if (!payload["id"]) {
    return { pass: false, reason: "missing id in payload", checks: {} };
  }

  const expectedIndex = loadExpected();
  const caseExpected = expectedIndex.get(caseId);
  if (!caseExpected) {
    return {
      pass: false,
      reason: `id ${JSON.stringify(caseId)} not in expected.json`,
      checks: {},
    };
  }

  const checks: Record<string, Check> = {};

  // Tool errors are never expected — always fail immediately.
  if (
    typeof candidate === "object" &&
    candidate !== null &&
    (candidate as Record<string, unknown>)["kind"] === "tool_error"
  ) {
    const c = candidate as Record<string, unknown>;
    checks["no_tool_error"] = {
      pass: false,
      detail: `tool crashed: ${c["name"]}: ${c["message"]}`,
    };
    return wrap(checks);
  }
  checks["no_tool_error"] = { pass: true, detail: "tool did not crash" };

  const expectedKind = caseExpected.expected.kind;

  if (expectedKind === "tagged") {
    checks["boundary_envelope"] = checkBoundaryEnvelope(candidate, caseExpected);
    return wrap(checks);
  }

  // Success path — candidate must be a plain object (record), not tagged.
  const candidateRecord = candidate as Record<string, unknown>;

  checks["shape"] = checkShape(candidateRecord);
  if (!checks["shape"]!.pass) return wrap(checks);

  checks["finite_value"] = checkFiniteValue(candidateRecord);
  if (!checks["finite_value"]!.pass) return wrap(checks);

  checks["method_admissible"] = checkMethodAdmissible(candidateRecord);

  const requestedPrecision = (inp as { precision?: number })["precision"] ?? 50;
  checks["self_reported_precision"] = checkSelfReportedPrecision(
    candidateRecord,
    requestedPrecision,
  );

  checks["value_accuracy"] = checkValueAccuracy(candidateRecord, caseExpected);

  return wrap(checks);
}

// ─── Main ─────────────────────────────────────────────────────────────────────

try {
  const payload = JSON.parse(readFileSync(0, "utf8")) as {
    id?: string;
    input?: Record<string, unknown>;
    candidate?: unknown;
  };
  const result = verify(payload);
  process.stdout.write(JSON.stringify(result) + "\n");
} catch (e) {
  process.stderr.write(String(e) + "\n");
  process.stdout.write(
    JSON.stringify({
      pass: false,
      reason: `verifier crashed: ${(e as Error).constructor?.name}: ${(e as Error).message}`,
      checks: {},
    }) + "\n",
  );
}

File: benchmarks/hypergeometric-pfq/golden/verify.py
#!/usr/bin/env python3
"""hypergeometric-pfq verifier — invariant-based, language-neutral.

stdin (one JSON object):
  {
    "input":     <input row from inputs.json>,
    "candidate": <candidate output — value/tagged/tool_error>,
    "id":        str
  }

  where `<candidate>` is one of:
    success record:
      {"value": {"re": "<dec>", "im": "<dec>"},
       "achieved_precision": int,
       "method": str,
       "n_terms": int,
       "working_precision": int,
       "warnings": [str, ...]}
    tagged boundary:
      {"kind": "tagged", "tag": "hypergeometric-pfq/<class>",
       "payload": {...}}
    tool error:
      {"kind": "tool_error", "name": str, "message": str}

stdout:
  {"pass": bool, "reason": str,
   "checks": {<name>: {"pass": bool, "detail": str}}}

Reads `expected.json` adjacent on disk for pinned truth values and
tolerance contracts.

Per ADR-0019 §1 the verifier checks INVARIANTS, not byte-equality.
Five invariants per success-path case:

  1. shape                  — required output fields present, types match
  2. finite_value           — re, im parse as finite mpmath floats
  3. value_accuracy         — relative error vs pinned truth ≤ tolerance
  4. method_admissible      — `method ∈ {closed-form-0F0, closed-form-1F0,
                                       direct-series}`
  5. self_reported_precision — `achieved_precision >= input.precision − 1`
                                (the tool may under-report by 1 dps;
                                full check delegated to value_accuracy)

Refusal-path cases get one check:

  6. boundary_envelope       — emitted output is `{kind: "tagged", tag: ...}`
                                with `tag == expected.expected.tag`

Tolerances are per-case in `expected.json::cases[i].tolerance_rel`
(decimal-string → mpmath.mpf). Tier-0 cases use `1e-(precision-2)`;
tier-C near-unit-circle cases relax to `1e-(precision-8)` because the
direct-series + cancellation-bump can underflow the requested
precision by a handful of bits in the slow-convergence regime.
"""

from __future__ import annotations

import json
import sys
import traceback
from pathlib import Path
from typing import Any, Optional

import mpmath
mpmath.mp.dps = 80  # truth values are pinned to ~80 dps; comparison budget

HERE = Path(__file__).resolve().parent
_EXPECTED_INDEX: Optional[dict[str, dict]] = None


def _load_expected() -> dict[str, dict]:
    """Load `expected.json` once and index by id."""
    global _EXPECTED_INDEX
    if _EXPECTED_INDEX is None:
        path = HERE / "expected.json"
        if not path.exists():
            _EXPECTED_INDEX = {}
            return _EXPECTED_INDEX
        payload = json.loads(path.read_text())
        _EXPECTED_INDEX = {c["id"]: c for c in payload["cases"]}
    return _EXPECTED_INDEX


# ---------------------------------------------------------------------
# Per-check helpers
# ---------------------------------------------------------------------

ADMITTED_METHODS = {"closed-form-0F0", "closed-form-1F0", "direct-series"}


def _is_str(x: Any) -> bool:
    return isinstance(x, str)


def _check_shape_value(candidate: dict) -> dict:
    """Verify the shape of a success record."""
    required = {
        "value", "achieved_precision", "method",
        "n_terms", "working_precision", "warnings",
    }
    missing = required - set(candidate.keys())
    if missing:
        return {"pass": False, "detail": f"missing fields: {sorted(missing)}"}

    v = candidate["value"]
    if not isinstance(v, dict) or set(v.keys()) != {"re", "im"}:
        return {"pass": False,
                "detail": f"value must be {{re, im}}; got {type(v).__name__} {v}"}
    if not (_is_str(v["re"]) and _is_str(v["im"])):
        return {"pass": False, "detail": "value.re and value.im must be strings"}

    if not isinstance(candidate["achieved_precision"], int):
        return {"pass": False, "detail": "achieved_precision must be int"}
    if not isinstance(candidate["n_terms"], int):
        return {"pass": False, "detail": "n_terms must be int"}
    if not isinstance(candidate["working_precision"], int):
        return {"pass": False, "detail": "working_precision must be int"}
    if not _is_str(candidate["method"]):
        return {"pass": False, "detail": "method must be str"}
    if not isinstance(candidate["warnings"], list) or not all(
        _is_str(w) for w in candidate["warnings"]):
        return {"pass": False, "detail": "warnings must be list[str]"}

    return {"pass": True, "detail": "all required fields present"}


def _check_finite_value(candidate: dict) -> dict:
    """Verify the candidate's re/im parse to finite mpmath numbers."""
    try:
        re = mpmath.mpf(candidate["value"]["re"])
        im = mpmath.mpf(candidate["value"]["im"])
    except Exception as e:
        return {"pass": False, "detail": f"could not parse value: {e}"}
    if not (mpmath.isfinite(re) and mpmath.isfinite(im)):
        return {"pass": False, "detail": f"non-finite value re={re}, im={im}"}
    return {"pass": True, "detail": f"value parses, finite"}


def _check_value_accuracy(candidate: dict, expected: dict) -> dict:
    """Compare candidate to pinned truth at the case's tolerance."""
    truth_re = mpmath.mpf(expected["truth"]["re"])
    truth_im = mpmath.mpf(expected["truth"]["im"])
    truth = mpmath.mpc(truth_re, truth_im)

    cand_re = mpmath.mpf(candidate["value"]["re"])
    cand_im = mpmath.mpf(candidate["value"]["im"])
    cand = mpmath.mpc(cand_re, cand_im)

    diff = mpmath.fabs(cand - truth)
    scale = mpmath.fabs(truth)
    if scale == 0:
        # Pinned truth is zero (rare); fall back to absolute.
        rel = float(diff)
    else:
        rel = float(diff / scale)

    return rel


def _check_method_admissible(candidate: dict) -> dict:
    m = candidate.get("method", "")
    if m not in ADMITTED_METHODS:
        return {"pass": False,
                "detail": f"method={m!r} not in admitted set {sorted(ADMITTED_METHODS)}"}
    return {"pass": True, "detail": f"method={m!r}"}


def _check_self_reported_precision(candidate: dict, requested_precision: int) -> dict:
    """The tool's `achieved_precision` field is its own precision report.

    Per ADR-0007 (agent-honest output), the tool may under-report by a
    small margin if cancellation eats into the requested precision; the
    bench's `value_accuracy` check is the authoritative gate. Here we
    verify only that `achieved_precision <= requested_precision` (the
    tool cannot over-report) and that the relationship is sane
    (non-negative).
    """
    ap = candidate["achieved_precision"]
    if ap < 0:
        return {"pass": False, "detail": f"achieved_precision={ap} is negative"}
    if ap > requested_precision:
        return {
            "pass": False,
            "detail": (f"achieved_precision={ap} exceeds requested precision "
                       f"{requested_precision} — tool over-reports"),
        }
    return {"pass": True,
            "detail": f"achieved={ap}, requested={requested_precision}"}


def _check_boundary_envelope(candidate: dict, expected: dict) -> dict:
    """Refusal-path cases: candidate must be tagged with the expected tag."""
    if not isinstance(candidate, dict) or candidate.get("kind") != "tagged":
        return {"pass": False,
                "detail": (f"expected tagged refusal but got "
                           f"{candidate.get('kind', type(candidate).__name__)}")}
    expected_tag = expected["tag"]
    actual_tag = candidate.get("tag", "")
    if actual_tag != expected_tag:
        return {"pass": False,
                "detail": f"tag {actual_tag!r} != expected {expected_tag!r}"}

    payload_predicate = expected.get("payload_predicate", {})
    payload = candidate.get("payload", {})
    reason_substr = payload_predicate.get("reason_substr")
    if reason_substr:
        reason = payload.get("reason", "")
        if reason_substr not in reason:
            return {"pass": False,
                    "detail": (f"payload.reason={reason!r} does not contain "
                               f"the predicate substring {reason_substr!r}")}
    return {"pass": True, "detail": f"tagged {expected_tag}"}


# ---------------------------------------------------------------------
# Top-level verify
# ---------------------------------------------------------------------

def verify(payload: dict) -> dict:
    case_id = payload.get("id", "")
    candidate = payload.get("candidate", {})
    inp = payload.get("input", {})

    if "id" not in payload:
        # The id is present on the corpus but not always on the wrapper;
        # some bench drivers don't include it. Look up by content if
        # possible — match the input's a/b/z/precision against expected.
        # For simplicity, require the id.
        return {"pass": False, "reason": "missing id in payload", "checks": {}}

    expected_index = _load_expected()
    if case_id not in expected_index:
        return {"pass": False,
                "reason": f"id {case_id!r} not in expected.json",
                "checks": {}}
    case_expected = expected_index[case_id]

    checks: dict[str, dict] = {}

    # Tool errors are never expected — refusals come back as tagged
    # envelopes, not as crashes.
    if isinstance(candidate, dict) and candidate.get("kind") == "tool_error":
        checks["no_tool_error"] = {
            "pass": False,
            "detail": (f"tool crashed: {candidate.get('name')}: "
                       f"{candidate.get('message')}"),
        }
        return _wrap(checks)
    checks["no_tool_error"] = {"pass": True, "detail": "tool did not crash"}

    expected = case_expected["expected"]
    expected_kind = expected.get("kind", "value")

    if expected_kind == "tagged":
        checks["boundary_envelope"] = _check_boundary_envelope(candidate, expected)
        return _wrap(checks)

    # Success path.
    checks["shape"] = _check_shape_value(candidate)
    if not checks["shape"]["pass"]:
        return _wrap(checks)

    checks["finite_value"] = _check_finite_value(candidate)
    if not checks["finite_value"]["pass"]:
        return _wrap(checks)

    checks["method_admissible"] = _check_method_admissible(candidate)

    requested_precision = inp["precision"]
    checks["self_reported_precision"] = _check_self_reported_precision(
        candidate, requested_precision)

    rel = _check_value_accuracy(candidate, expected)
    tol = mpmath.mpf(case_expected["tolerance_rel"])
    if rel <= float(tol):
        checks["value_accuracy"] = {
            "pass": True,
            "detail": f"rel={rel:.3e} ≤ tol={float(tol):.3e}",
        }
    else:
        checks["value_accuracy"] = {
            "pass": False,
            "detail": f"rel={rel:.3e} > tol={float(tol):.3e}",
        }

    return _wrap(checks)


def _wrap(checks: dict) -> dict:
    overall = all(c["pass"] for c in checks.values())
    if overall:
        return {"pass": True, "reason": "all invariants hold", "checks": checks}
    first_fail = next(k for k, v in checks.items() if not v["pass"])
    return {"pass": False,
            "reason": f"failed: {first_fail} — {checks[first_fail]['detail']}",
            "checks": checks}


def main() -> None:
    try:
        payload = json.load(sys.stdin)
        result = verify(payload)
    except Exception as e:
        sys.stderr.write(traceback.format_exc())
        sys.stderr.write("\n")
        result = {
            "pass": False,
            "reason": f"verifier crashed: {type(e).__name__}: {e}",
            "checks": {},
        }
    json.dump(result, sys.stdout)
    sys.stdout.write("\n")


if __name__ == "__main__":
    main()
